### PR TITLE
*: mini refactor

### LINKED ISSUE #4739: interpreter planner refactor
#### background
Currently tiflash's interpreter is designed based on `DAGQueryBlock`.
`DAGQueryBlock` is designed to coprocess read, and it can handle simple queries well.
However, there is no way to handle the complex query of MPP, so the `DAGQueryBlock` tree is introduced, which is a compromise that makes the code more complicated and affects the subsequent iterative development.
We need to refactor the interpreter code in a plan visit way.

#### design
Therefore we introduce a new abstraction layer called `PhysicalPlan` to replace `DAGQueryBlock`.
It will be responsible for the processing of `ExpressionAction` and the generation of `BlockInputStream`.
We call this organization of code the planner.

```
DAGRequest (executor tree or executor array).
     |
     |
     v
physical plan <---+
     |            |
     |            | optimize
     |            |
     +------------+   
     |
     |
     v
BlockInputStream
```

#### why planner
Most databases are designed based on planner. 
- Friendly for database developers to understand code.
- We can improve tiflash with a general approach from the DB area, such as pipeline mode.

#### todo list
- [x] public desgin doc
    - #4743
- [x] Small refactoring of Interpreter before planner refactoring 
     - #4118 
- [x] interpreter test framework 
    - #4609 
- [x] planner refactory
    - [x] poc #4601 
    - [x] Introduce `PhysicalPlan`/`planner`, add switch and set it to planner by default
        - https://github.com/pingcap/tiflash/pull/4820 
    - [x] Replace all operators with `PhysicalPlan`
        - [x] common plans: projection/agg/filter/exchange_(sender/receiver)/limit/topn
            - https://github.com/pingcap/tiflash/pull/5143
        - [x] window function
            - https://github.com/pingcap/tiflash/pull/5363
        - [x] join
            - https://github.com/pingcap/tiflash/pull/5320
        - [x] table scan
            - https://github.com/pingcap/tiflash/pull/5335
    - [x] remove `DAGQueryBlock`
            - https://github.com/pingcap/tiflash/pull/5381
- [x] merge to master
    - feature branch: https://github.com/pingcap/tiflash/tree/planner_refactory
    - https://github.com/pingcap/tiflash/pull/5493
    - https://github.com/pingcap/tiflash/pull/5321

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #4739 

Problem Summary:

### What is changed and how it works?

port from https://github.com/pingcap/tiflash/pull/5143

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Flash/tests/gtest_interpreter.cpp [hunk 1]
```diff
@@ -222,15 +208,12 @@ Union: <for test>
 Union: <for test>
  Expression x 10: <final projection>
   Expression: <projection>
-   Expression: <before projection>
-    Expression: <final projection>
-     Expression: <projection>
-      Expression: <before projection>
+   Expression: <final projection>
```
Comment (ywqzzy) [inline dbms/src/Flash/tests/gtest_interpreter.cpp:211]: Maybe clarify why the “before projection expression” is deleted

File: dbms/src/Interpreters/Settings.h [hunk 2]
```diff
@@ -209,7 +209,7 @@ struct Settings
       * Basically, limits are checked for each block (not every row). That is, the limits can be slightly violated. \
       * Almost all limits apply only to SELECTs. \
       * Almost all limits apply to each stream individually. \
-      */                                                                                                                                                                       \
```
Comment (ywqzzy) [inline dbms/src/Interpreters/Settings.h:212]: Why the file changed？

File: dbms/src/Flash/tests/gtest_interpreter.cpp (hunk 1 above)
Comment (SeaRise) [reply dbms/src/Flash/tests/gtest_interpreter.cpp:211]: due to https://github.com/pingcap/tiflash/pull/5326/files#diff-c21f466589627c568489360b256d315d80d6bf114ad7706fffc561183952172cR49

File: dbms/src/Interpreters/Settings.h (hunk 2 above)
Comment (SeaRise) [reply dbms/src/Interpreters/Settings.h:212]: just format the '\'

**Review by windtalker — APPROVED** [review]
LGTM

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**